doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -5,269 +5,297 @@ Reasoning with inductive types
-      and performs case analysis using this subterm.
+      apply or_introl.  (* equivalent to "left" *)
 
-   .. tacv:: destruct {+, @term}
 
-      This is a shortcut for :n:`destruct @term; ...; destruct @term`.
+.. _CaseAnalysisAndInduction:
 
-   .. tacv:: destruct @term as @or_and_intropattern_loc
+Case analysis
+-------------
 
-      This behaves as :n:`destruct @term` but uses the names
-      in :token:`or_and_intropattern_loc` to name the variables introduced in the
-      context. The :token:`or_and_intropattern_loc` must have the
-      form :n:`[p11 ... p1n | ... | pm1 ... pmn ]` with ``m`` being the
-      number of constructors of the type of :token:`term`. Each variable
-      introduced by :tacn:`destruct` in the context of the ``i``-th goal
-      gets its name from the list :n:`pi1 ... pin` in order. If there are not
-      enough names, :tacn:`destruct` invents names for the remaining variables
-      to introduce. More generally, the :n:`pij` can be any introduction
-      pattern (see :tacn:`intros`). This provides a concise notation for
-      chaining destruction of a hypothesis.
+The tactics in this section implement case
+analysis on inductive or co-inductive objects (see :ref:`gallina-inductive-definitions`).
 
-   .. tacv:: destruct @term eqn:@naming_intropattern
-      :name: destruct … eqn:
+Some useful terms:
 
-      This behaves as :n:`destruct @term` but adds an equation
-      between :token:`term` and the value that it takes in each of the
-      possible cases. The name of the equation is specified
-      by :token:`naming_intropattern` (see :tacn:`intros`),
-      in particular ``?`` can be used to let Coq generate a fresh name.
+* `n` is a :gdef:`dependent premise` of `forall n:nat, n + 0 = n` because
+  `n` appears both in the binder of the `forall` and in the quantified statement
+  `n + 0 = n`.
 
-   .. tacv:: destruct @term with @bindings
+* `A` and `B` are :gdef:`non-dependent premises <non-dependent premise>` of `A -> B -> C`
+  because they don't appear in a `forall` binder.
+  (Recall that `A -> B` is notation for the term `forall _ : A, B`; the wildcard
+  `_` can't be referred to in the quantified statement.)
 
-      This behaves like :n:`destruct @term` providing explicit instances for
-      the dependent premises of the type of :token:`term`.
+.. tacn:: destruct @induction_clause_list
 
-   .. tacv:: edestruct @term
-      :name: edestruct
+   .. insertprodn induction_clause_list as_or_and_ipat
 
-      This tactic behaves like :n:`destruct @term` except that it does not
-      fail if the instance of a dependent premises of the type
-      of :token:`term` is not inferable. Instead, the unresolved instances
-      are left as existential variables to be inferred later, in the same way
-      as :tacn:`eapply` does.
+   .. prodn::
+      induction_clause_list ::= {+, @induction_clause } {? using @one_term {? with @bindings } } {? @occurrences }
```

Comment:
To make things more readable, I suggest:

1. Move the trailing `{? @occurrences }` inside the optional `using` clause. When the `using` clause` is not provided, there is anyway some ambiguity wrt which parsing path is selected in the current presentation.
2. Introduce a new nonterminal for `using @one_term {? with @bindings } {? @occurrences }`.
3. Now that the definition of `induction_clause_list` is much shorter (reduced to `{+, @induction_clause } {? induction_principle }`), it becomes reasonable to splice it.

Reply:
What's a better name for `destruction_arg`?

Reply:
![image](https://user-images.githubusercontent.com/1253341/122153210-5de83200-ce17-11eb-8c2e-668bd87c9875.png)

Reply:
I don't have great ideas. This could become `term_or_position_argument`. Or maybe it would make sense to splice it and simply replace it everywhere with `{| @natural | @one_term_with_bindings }`...

Reply:
`induction_arg` would be a little better, at least similar to the other NT names seen here.